From 4db9e30bce4b0bfe14c9095195c660595622473a Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Mon, 6 Feb 2023 19:36:03 +0530
Subject: [PATCH] cfg80211/mac80211: add MLO support for HE-BSS color change
 events

Currently, HE-BSS color change event is handled only for non-MLO case.

Add support for HE-BSS color change event with MLO.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c         |  10 +-
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1667,16 +1667,17 @@ void ath12k_mac_bcn_tx_event(struct ath1
 	if (!link_conf->color_change_active && !arvif->bcca_zero_sent)
 		return;
 
-	if (link_conf->color_change_active && ieee80211_beacon_cntdwn_is_complete(vif)) {
+	if (link_conf->color_change_active &&
+	    ieee80211_beacon_cntdwn_is_complete(vif, arvif->link_id)) {
 		arvif->bcca_zero_sent = true;
-		ieee80211_color_change_finish(vif);
+		ieee80211_color_change_finish(vif, arvif->link_id);
 		return;
 	}
 
 	arvif->bcca_zero_sent = false;
 
 	if (link_conf->color_change_active && !vif->bss_conf.ema_ap)
-		ieee80211_beacon_update_cntdwn(vif, 0);
+		ieee80211_beacon_update_cntdwn(vif, arvif->link_id);
 	ieee80211_queue_work(ar->ah->hw, &arvif->update_bcn_template_work);
 }
 
@@ -9044,7 +9045,8 @@ static void ath12k_update_obss_color_not
 	if (arvif->is_created)
 		ieeee80211_obss_color_collision_notify(arvif->ahvif->vif,
 						       arvif->obss_color_bitmap,
-						       GFP_KERNEL);
+						       GFP_KERNEL,
+						       arvif->link_id);
 	arvif->obss_color_bitmap = 0;
 	mutex_unlock(&ar->conf_mutex);
 }
